rts/lra/lra-test/lra-test-basic/src/test/java/io/narayana/lra/arquillian/resource/FailingAfterLRAListener.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2019, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package io.narayana.lra.arquillian.resource;
+
+import org.eclipse.microprofile.lra.annotation.AfterLRA;
+import org.eclipse.microprofile.lra.annotation.ws.rs.LRA;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Path(FailingAfterLRAListener.ROOT_PATH)
+public class FailingAfterLRAListener {
+
+    public static final String ROOT_PATH = "failing-after-lra-listener";
+    public static final String ACTION_PATH = "action";
+
+    private static final AtomicInteger afterLRACounter = new AtomicInteger(0);
+
+    @GET
+    @Path(ACTION_PATH)
+    @LRA(LRA.Type.REQUIRED)
+    public String doWorkInLRA(@HeaderParam(LRA.LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return lraId.toASCIIString();
+    }
+
+    @PUT
+    @Path("after-lra")
+    @AfterLRA
+    public Response afterLRA(@HeaderParam(LRA.LRA_HTTP_ENDED_CONTEXT_HEADER) URI endedLRAId) {
+        if (afterLRACounter.getAndIncrement() < 2) {
```

Comment:
@mmusgrov please can you add a reference to https://issues.redhat.com/browse/JBTM-3163 here so we know why it is doing this twice. IIUC the [comment](https://github.com/jbosstm/narayana/pull/1559#issuecomment-582071967) from @xstefank then if that is fixed then [this](https://github.com/jbosstm/narayana/pull/1559/files#diff-f1f5ec6c4099230d8d191d68c62295cdR100) line will probably start failing as afterLRA will only be called twice.

Reply:
@tomjenkinson no, it shouldn't start failing. The recovery will just be replayed two times.

Reply:
If I understand you correctly, once https://issues.redhat.com/browse/JBTM-3163 is fixed instead of two notifications for one recovery we could go down to "getAndIncrement < 1"

Am I right in thinking therefore that https://github.com/jbosstm/narayana/pull/1559/files#diff-6ba4dd5973413b54d0299d8935a156c1R54-R56 is not actually required or used in this PR? If so I think we should remove that part? I will create a discussion localised on that line for this.

Reply:
Actually I see the code may well apply if the scan is in progress

Reply:
It is possible to move it to < 1 after https://issues.redhat.com/browse/JBTM-3163 is resolved, but not required. 

Sorry, the second link doesn't probably take me where you've intended. But if it's NarayanaLRARecovery then that code is correct. The two invocations there are about something different -- namely the clash with periodic recovery which may sometime happen. That is a different issue which was decided that it will not be resolved some time ago.

Reply:
I understand this better now. Currently we need 3 total calls to this method  (two from the normal cleanup, one (though will receive 2) from recovery)

In the future we need 2 notifications (one from the normal cleanup, one from recovery) so will be able to change this line to `if (afterLRACounter.getAndIncrement() < 1)` after JBTM-2863 is solved. As such it needs the comment and a reference to the bit that needs cleanup adding to the JBTM

Reply:
I am not clear on what this discussion is about. The test says `afterLRACounter.getAndIncrement() < 2` so it will return a 500 error status 2 times followed by a 200 status. The test is verifying that the LRA implementation will keep attempting the AfterLRA notification until it receives an HTTP 200 OK response. So in all we expect 2+1=3 notifications. The test FailingParticipantCallsIT asserts that the callback was invoked 3 times.

On the issue of JBTM-3163 that should have been incidentally resolved by the PR but it seems irrelevant to this PR.

Reply:
The discussion is fundamentally related to the "2" in your formula "2+1=3", in other words why we have the need to return the ISE twice rather than just once.

The reason I understand that the test originally needed to skip "2" calls before was because of a duplicate notification (JBTM-3163) so the two calls would have been from the same routine execution in Narayana ([this](https://issues.redhat.com/browse/JBTM-3163?focusedCommentId=13764034&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13764034)?) and so returning OK to the second notification would not have tested the call was repeated.

The premise of the observation is related to why the test not just say:
```
if (afterLRACounter.getAndIncrement() < 1) {
     return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
}
return Response.ok().build();
```

Reply:
@mmusgrov I think this is the last area to get clarity on. Do we need to return ISE on the first 2 calls to `afterLRA` or could it just be one now? And what is the plan to resolve JBTM-3163, will it be linked to this PR somehow?

Reply:
Well okay I changed it to 2 to get faster test turnaround time on testing.